src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -815,6 +793,32 @@ public struct Enumerator: IEnumerator<KeyValuePair<TKey,TValue>>,
             }
         }
 
+        // This is separated into a new class to save on memory by reducing
+        // fields if Keys/Values are never called.
+        // Since it shouldn't be visible to the outside world, it also serves
+        // as the SyncRoot (also rarely used) which saves an additional field.
+        // The downside of this is that Keys/Values are called, then there will
+        // be an additional object allocation the first time, plus this class
+        // needs to be initialized with Interlocked.CompareExchange.
+        private sealed class KeysAndValuesHolder
+        {
+            private readonly Dictionary<TKey, TValue> dictionary;
+            private KeyCollection keys;
+            private ValueCollection values;
+
+            public KeysAndValuesHolder(Dictionary<TKey, TValue> dictionary)
+            {
+                Contract.Assert(dictionary != null);
+                this.dictionary = dictionary;
+            }
+
+            public KeyCollection Keys =>
+                keys ?? (keys = new KeyCollection(dictionary));
```

Comment:
What's even more bizarre, doing this:

```cs
public KeyCollection Keys
{
    [MethodImpl(AggressiveInlining)]
    get
    {
        return keys ?? InitializeKeys();
    }
}

[MethodImpl(NoInlining)]
private KeyCollection InitializeKeys()
{
    keys = new KeyCollection(dictionary);
    return keys;
}
```

seems to kill inlining of *both* `KeysAndValuesHolder.get_Keys` and `Dictionary.get_KeysAndValues`. No idea why.